java/src/main/kotlin/io/spine/validation/java/DistinctFieldGenerator.kt

```diff
@@ -0,0 +1,148 @@
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.validation.java
+
+import io.spine.base.FieldPath
+import io.spine.protodata.ast.isList
+import io.spine.protodata.ast.isMap
+import io.spine.protodata.ast.name
+import io.spine.protodata.ast.qualifiedName
+import io.spine.protodata.java.CodeBlock
+import io.spine.protodata.java.Expression
+import io.spine.protodata.java.ReadVar
+import io.spine.protodata.java.StringLiteral
+import io.spine.protodata.java.call
+import io.spine.protodata.java.field
+import io.spine.protodata.java.toBuilder
+import io.spine.validate.ConstraintViolation
+import io.spine.validation.DistinctField
+import io.spine.validation.PATTERN
+import io.spine.validation.java.ErrorPlaceholder.FIELD_DUPLICATES
+import io.spine.validation.java.ErrorPlaceholder.FIELD_PATH
+import io.spine.validation.java.ErrorPlaceholder.FIELD_TYPE
+import io.spine.validation.java.ErrorPlaceholder.FIELD_VALUE
+import io.spine.validation.java.ErrorPlaceholder.PARENT_TYPE
+import io.spine.validation.java.ValidationCodeInjector.MessageScope.message
+import io.spine.validation.java.ValidationCodeInjector.ValidateScope.parentPath
+
+/**
+ * The generator for `(distinct)` option.
+ *
+ * Generates code for a single field represented by the provided [view].
+ */
+internal class DistinctFieldGenerator(private val view: DistinctField) {
+
+    private val field = view.subject
+    private val fieldType = field.type
+    private val getter = message.field(field).getter<List<*>>()
+    private val declaringType = field.declaringType
+
+    /**
+     * Generates code for a field represented by the [view].
+     */
+    fun generate(): FieldOptionCode {
+        val collection = validatedCollection()
+        val set = ImmutableSetClass.call<Set<*>>("copyOf", collection)
+        val constraint = CodeBlock(
+            """
+            if (!$collection.isEmpty() && $collection.size() != $set.size()) {
+                var duplicates = ${extractDuplicates(collection)};
+                var fieldPath = ${fieldPath(parentPath)};
+                var violation = ${violation(ReadVar("fieldPath"), ReadVar("duplicates"))};
+                violations.add(violation);
+            }
+            """.trimIndent()
+        )
+        return FieldOptionCode(constraint)
+    }
+
+    /**
+     * Returns an expression containing the collection that should not have duplicate values.
+     *
+     * The resulting expression depends on the [fieldType]:
+     *
+     * 1. For `repeated` fields, it is a field value itself.
+     * 2. For `map` fields, it is a collection of map values.
+     */
+    private fun validatedCollection(): Expression<Collection<*>> = when {
+        fieldType.isList -> getter
+        fieldType.isMap -> getter.call("values")
+        else -> error(
```

Comment:
Is it error or compilation error? If it's compilation error the language should not refer to `DistinctFieldGenerator` or other internals. It should talk to the user in terms of types and options he sees.

Reply:
This is an error. Field type checks are done by policies.